Configure own schema with explicit id

jgrocha opened the pull request.

marcjansen commented:
Hey @jgrocha, in #112 you say that without this, it is impossible to extend the model, right? This cries out for a test that fails with current master and passes with the changes included here. 

Do you think you can write a test confirming that with this change one can extend the model? If not, someone else can do this at a later point.

jgrocha commented:
Hi Marc,
It makes some sense to include a test to check that our schema can coexist with other schemas. But in this specific case, the error results from a ES6 limitation that imposes unique ids when working with more than one schema in the same application. Nothing is said about that in the [schema documentation](https://docs.sencha.com/extjs/6.0/6.0.0-classic/#!/api/Ext.data.schema.Schema), unfortunately. It would make more sense to add the test to ES6.

marcjansen commented:
I only partly agree.

**Yes**: The test would make sense in ExtJS6 (and it probably already exists).

But… also **no**: since we as a library limit the ability of users of the library (because we don't set the `id`), I think it is our responsibility to ensure that this won't happen again, e.g. in future refactorings.

If I see this thin in say 3 months, I may be tempted to remove the hardcoded id … and the bug would reappear. But not if there is a test that fails and warns me.

That said, if you cannot (or don't want) to create a test, that is fine for me and this PR, it simply means that I (or someone else) probably should do it.

Does that sound reasonable?

jgrocha commented:
I've added a test to check if a unique `id` was assigned to our `GeoExt.data.model.Base` schema.

The reasoning behind: If we define a schema without an explicit `id`, it gets the value `"default"`. That's the ES6 problem: if more than one schema is present without an explicit `id`, both get the same `"default"` value.
The test I added guarantees that there is an `id` and it is not `"default"`.

Please double check if this sounds good enough.

marcjansen commented:
Thanks @jgrocha. Just to make extra sure I understand the root of the problem:

If I go to [the current map-example](http://rawgit.com/geoext/geoext3/master/examples/component/map.html), open a console, and enter:

```js
Ext.define('TestModel', {
    extend: 'Ext.data.Model',
    schema: {
        namespace: 'TestModel'
    }
});
```

… this should trigger the error mentioned in #112, right?

Yet I fail to reporoduce this. What are the steps needed to reproduce the error?

Sorry that such a trivial change takes that much back-and-forth.

jgrocha commented:
You are right @marcjansen. I did not documented the error well enough.
The error can be reproduced in this simple [fiddle](https://fiddle.sencha.com/#fiddle/12q6). 
You can not reproduce the error with a built version of ExtJS. That's why you can't get it on a running example. But if you run any ExtJS 6 project in debug mode (from the sources), the error is triggered. So **this error happens in debug mode, when you have a project with more than one schema defined**.
It does not have nothing to do with GeoExt 3, but since we define a schema for our data classes, we might want to add an `id`. We can keep the code as it is and let developers decided what to do if they catch the error.

marcjansen commented:
OK. I now got it. Since the relevant code in ExtJS is wrapped in a `<debug>`-tag, we only see the error in the debug build. For reference, this method throws: http://docs.sencha.com/extjs/6.0/6.0.0-classic/source/Schema.html#Ext-data-schema-Schema-static-method-get

This is good to go then. Thanks @jgrocha.
